src/connection.rs

```diff
@@ -48,21 +54,101 @@ impl AsyncBufRead for Connection {
         }
     }
 
     fn consume(self: Pin<&mut Self>, amt: usize) {
         match self.project() {
             ConnectionProjection::Tcp(s) => s.consume(amt),
+            ConnectionProjection::Unix(s) => s.consume(amt),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum Addr {
+    Tcp(String),
+    Unix(String),
+    Unknown(String),
+}
+
+impl Addr {
+    const DEFAULT_PORT: u16 = 11211;
+
+    fn parse(dsn: &str) -> Result<Self, Error> {
+        let url = url::Url::parse(dsn).map_err(|e| {
+            Error::Connect(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("failed to parse DSN: {}", e),
+            ))
+        })?;
+
+        match url.scheme() {
+            "unix" => Ok(Addr::Unix(url.path().to_string())),
+            "tcp" => Ok(Addr::Tcp(format!(
+                "{}:{}",
+                url.host_str().ok_or_else(|| {
+                    Error::Connect(io::Error::new(
+                        io::ErrorKind::InvalidInput,
+                        "no host found in DSN",
+                    ))
+                })?,
+                url.port().unwrap_or(Self::DEFAULT_PORT)
+            ))),
+            _ => Ok(Addr::Unknown(dsn.to_string())),
         }
     }
 }
 
 impl Connection {
-    pub async fn new<S: AsRef<str>>(dsn: S) -> Result<Connection, Error> {
-        TcpStream::connect(dsn.as_ref())
-            .await
-            .map(|c| Connection::Tcp(BufReader::new(BufWriter::new(c))))
-            .map_err(Error::Io)
+    pub async fn new<S: AsRef<str>>(dsn: S) -> Result<Self, Error> {
+        match Addr::parse(dsn.as_ref())? {
+            Addr::Unix(path) => UnixStream::connect(path)
+                .await
+                .map(|c| Connection::Unix(BufReader::new(BufWriter::new(c))))
+                .map_err(Error::Connect),
+            Addr::Tcp(url) | Addr::Unknown(url) => TcpStream::connect(url)
+                .await
+                .map(|c| Connection::Tcp(BufReader::new(BufWriter::new(c))))
+                .map_err(Error::Connect),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Addr, Error};
+
+    #[tokio::test]
+    async fn test_unknown_scheme() {
+        assert_eq!(
+            Addr::parse("localhost:11211").unwrap(),
+            Addr::Unknown("localhost:11211".to_string())
+        )
+    }
+
+    #[tokio::test]
+    async fn test_tcp_scheme() {
+        assert_eq!(
+            Addr::parse("tcp://localhost:11211").unwrap(),
+            Addr::Tcp("localhost:11211".to_string())
+        )
+    }
+
+    #[tokio::test]
+    async fn test_unix_scheme() {
+        assert_eq!(
+            Addr::parse("unix:///tmp/memcached.sock").unwrap(),
+            Addr::Unix("/tmp/memcached.sock".to_string())
+        )
+    }
+
+    #[tokio::test]
+    async fn test_invalid_url() {
+        assert!(matches!(
+            Addr::parse("tcp://").unwrap_err(),
+            Error::Connect(_)
+        ));
     }
```

Comment:
Is it possible to avoid the unwraps and include the result in the `assert_eq!`? Like `assert_eq!(Addr::parse("localhost:11211"), Ok(Addr::Unknown("localhost:11211".to_string())))`?